http-client/Network/HTTP/Client/Request.hs

```diff
@@ -121,27 +120,21 @@ getUri req = URI
 
 applyAnyUriBasedAuth :: URI -> Request -> Request
 applyAnyUriBasedAuth uri req =
-    if hasAuth
-      then applyBasicAuth (S8.pack theuser) (S8.pack thepass) req
-      else req
+    case extractBasicAuthInfo uri of
+        Just auth -> uncurry applyBasicAuth auth req
+        Nothing -> req
+
+-- | Extract basic access authentication info in URI.
+-- Return Nothing when there is no auth info in URI.
+extractBasicAuthInfo :: URI -> Maybe (S8.ByteString, S8.ByteString)
+extractBasicAuthInfo uri = do
+    userInfo <- uriUserInfo <$> uriAuthority uri
+    when (':' `notElem` userInfo) Nothing
+    let username = takeWhile (/=':') userInfo
```

Comment:
It would be more efficient to use `let (username', password') = break (== ':') userInfo`